### ISSUE TITLE
A way to run context-menu items

### ISSUE THREAD

**Issue by arigo** [SketchUp/api-issue-tracker#425]
# Feature Request

In SketchUp (up to 2020 at least) the Ruby API is missing a way to execute context-menu items.  More precisely: if I right-click on, say, a component, I get many menu items that perform specific actions on this component.  Equivalently, I can select the component and then the "Edit" menu ends in a submenu called, say, "Component (1 in model)", with the same menu items.  Some of these items are available in the general API, like say ``#explode``.  Some of them are not, like say "Reload" (or "Edit Component" before SU2020).  The same issue exists not only about components but more generally all types of entities.

This is almost a case where we can use ``Sketchup.send_action(id)``: there is a long tail of very rarely useful commands that some person somewhere might want to call once.  It's certainly not worth adding a specific API for every last one of these actions.  But the problem is that the id of these context-menu items is variable.  I think that internally SketchUp builds the context menu and assigns ids dynamically every time it does.

A possible fix would be for SketchUp to use a repeatable way to assign ids to these commands.  Then the ids could be listed in http://ruby.sketchup.com/Sketchup.html#send_action-class_method .  Alternatively, give all these commands a name available in ``Sketchup.send_action(:foo)``.

(Note that it is already a problem that the ids listed numerically on this page are only available on the PC version.  Maybe a proper fix for this issue would also include supporting all of them on Mac too, maybe by giving them all names.  Though I guess the problem is that you don't want to guarantee that all these commands will continue to exist with no refactoring in future versions of SketchUp.  Maybe still give command names but be clear that these might be removed or replaced in future versions?)

**Comment by Eneroth3** [comment]
I suspect this lies outside the scope of the API. The API generally has low level functionality that can be combined to make (in theory) everything. It doesn't focus on implementing hooks for what exists in the UI. Already now the Windows keys for send_action lies outside of the API contract.

> (Note that these are officially “unsupported” and are not guaranteed to work in current or future versions of the API.)

Also referencing the UI directly would be quite fragile as the UI can change. What if some entry is removed, or renamed or in other ways changed?

I think it's better to open issues for the low level functionality needed, e.g. reloading a definition from a given file (#50), than looking too much at the details in the UI. At the top of my head I can't come up with more examples, but there are probably a few.

Please note that API design isn't just about getting the job done here and now, but about building a good foundation for the future.

**Comment by arigo** [comment]
OK, agreed.  I think I'll close this and reopen specific issues for specific missing features if more show up for me.

**Comment by DanRathbun** [comment]
I've only found a few items on the context menus that are missing from the API.
I filed one for the "Create group from Slice" context item already.

> But the problem is that the id of these context-menu items is variable. I think that internally SketchUp builds the context menu and assigns ids dynamically every time it does.

This is not unique to SketchUp. It's a inherent behavior of the way that [MFC](https://docs.microsoft.com/en-us/cpp/mfc/reference/ccontextmenumanager-class?view=vs-2019) builds context menus. Ie, it wants to reuse IDs. I guess it helps keep ID count down.
